src/lib.rs

```diff
@@ -1136,14 +1147,73 @@ impl<T> Drop for Async<T> {
     }
 }
 
-impl<T: Read> AsyncRead for Async<T> {
+/// Types whose I/O trait implementations do not drop the underlying I/O source.
+///
+/// The resource contained inside of the [`Async`] cannot be invalidated. This invalidation can
+/// happen if the inner resource (the [`TcpStream`], [`UnixListener`] or other `T`) is moved out
+/// and dropped before the [`Async`]. Because of this, functions that grant mutable access to
+/// the inner type are unsafe, as there is no way to guarantee that the source won't be dropped
+/// and a dangling pointer won't be left behind.
+///
+/// Unfortunately this extends to implementations of [`Read`] and [`Write`]. Since methods on those
+/// traits take `&mut`, there is no guarantee that the implementor of those traits won't move the
+/// source out while the method is being run.
+///
+/// This trait is an antidote to this predicament. By implementing this trait, it is guaranteed
+/// that using any I/O traits won't desroy the source. This way, [`Async`] can implement the
+/// `async` version of these I/O traits, like [`AsyncRead`], [`AsyncWrite`] and [`AsyncSeek`].
+///
+/// # Safety
+///
+/// Any I/O trait implementations for this type must not drop the underlying I/O source. Traits
+/// affected by this trait include [`Read`], [`Write`] and [`Seek`].
+///
+/// This trait is implemented by default on top of `libstd` types. In addition, it is implemented
+/// for immutable reference types, as it is impossible to invalidate any outstanding references
+/// while holding an immutable reference, even with interior mutability. As Rust's current pinning
+/// system relies on similar guarantees, I believe that this approach is robust.
+///
+/// [`Read`]: https://doc.rust-lang.org/std/io/trait.Read.html
+/// [`Seek`]: https://doc.rust-lang.org/std/io/trait.Seek.html
+/// [`Write`]: https://doc.rust-lang.org/std/io/trait.Write.html
+///
+/// [`AsyncRead`]: https://docs.rs/futures-io/latest/futures_io/trait.AsyncRead.html
+/// [`AsyncSeek`]: https://docs.rs/futures-io/latest/futures_io/trait.AsyncSeek.html
+/// [`AsyncWrite`]: https://docs.rs/futures-io/latest/futures_io/trait.AsyncWrite.html
+pub unsafe trait IoSafe {}
```

Comment:
I may be saying something silly because I haven't thought too much about it:

- All functions we call within the mut operation are safe and safe functions cannot violate the soundness of other safe code.
- The safety requirement for the `BorrowedFd` returned by `as_fd` is that the fd is not invalid while `BorrowedFd` is alive; if the fd is not valid when `as_fd` is called, it cannot create a `BorrowedFd` to return. (Perhaps they need to panic.)
- If the fd returned before and after the mut operation is different, the original fd may have been replaced. The fd we registered may already be invalid.

Assuming that the fd cannot be concurrently accessed by another thread during the mut operation, couldn't safety here be handled by making sure that the fd returned by `as_fd().as_raw_fd()` matches before and after the mut operation? (I guess we also need a guard to handle that in case the fd is invalidated, though.)

Something like:

```rs
let fd = self.get_ref().as_fd().as_raw_fd();

// mut operation...

let guard = ... // 
assert_eq!(fd, self.get_ref().as_fd().as_raw_fd());
mem::forget(guard);
```

Reply:
That's a good idea, but unfortunately I don't think it works out. The main issue is what to do if it turns out that the file descriptor is different.

- You can't `delete` the source from the `Poller`. In the time after the original source was dropped, a new source could be created (potentially in another thread) with the same FD and registered into the `Poller`. Doing this deregisters the new source from the `Poller` and results in hard to track down bugs. While it's not *technically* undefined behavior, it's not robust either.
- At the same time, you can't *not* `delete` the source from the `Poller`. Let's say the file descriptor slot hasn't been reallocated yet. Once the `async-io` thread runs `poller.poll()` with a dangling handle it immediately invokes unsound and undefined behavior. Practically this is an issue for Windows and `poll()`-syscall backends, as they both maintain user-space lists of poll-able handles and deleting them like this can lead to random errors or, worse, memory corruption. It has also been observed to be an issue in `kqueue` for some BSDs. Real life effects can include random spurious events and `Waker`s never being woken.

Essentially we're forced into a [Catch-22](https://en.m.wikipedia.org/wiki/Catch-22_(logic)) where, if we're unsure of the current fate of the file descriptor, the only available options lead to buggy and unsound code. Therefore we have to make it an invariant of the `Async` type that the file descriptor is alive at all times.

Reply:
well, then panic! if that happens (because we can assume that we can't recover the `Poller` easily)

Reply:
We can't `panic!` because other threads (and even the current thread) can keep going after a panic, which would result in even more undefined behavior. The proper thing to do in this case would be to `abort()`. However I would like to avoid aborting if possible, as that brings down the entire program which is disastrous for long running programs like web servers. I think we would be better citizens of the Rust ecosystem if we just make this precondition unsafe rather than trying to paper over the unsafety with panics or aborts.

Reply:
well, we could `abort()` when build in debug mode...

Reply:
I guess we could...

Reply:
@notgull 

Thanks for the explanation. IIUC, it appears that the bad cases you mention here is that it can be reproduced even if you wrap the underlying Io object with a Mutex or something.

Reply:
Nah, if using Mutex or something, the problem may not really occur with safe code because of the guard's lifetime issue.

Reply:
> well, we could `abort()` when built in debug mode...

This would require some code restructuring that I think should be reserved for a future PR. I'll resolve this convo for now.